### PR TITLE
[`Quantization`] Add str to enum conversion for AWQ

### LINKED ISSUE #27318: AWQ model error when `AwqConfig['version']` is a string
### System Info

A100 Cuda 12.2 transformers 4.36.0.dev (Nov4)

### Who can help?

@younesbelkada @SunMarc 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

load some AWQ model, like `"marcsun13/Llama-2-13B-AWQ"` - it has  `AwqConfig['version']` as a string "GEMM".
later in `utils/quantization.config.py::AWQcofig.post_init` there is this check:
```
        if self.version not in [AWQLinearVersion.GEMM, AWQLinearVersion.GEMV]:
            raise ValueError(
                f"Only supported versions are in [AWQLinearVersion.GEMM, AWQLinearVersion.GEMV] - not recognized version {self.version}"
            )
```
Apparently it fails because `version` is still a string "GEMM"

please consider some string -> enum conversion during init or rewrite this check when loading model.

### Expected behavior

normal model loading.

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/27318

Some repositories that use the old conversion script have the `version` field set to `GEMM` whereas it should be `gemm`. Despite the fix should be that we should convert the config file accordingly e.g.: https://huggingface.co/marcsun13/Llama-2-13B-AWQ/discussions/1 I propose to add a small str to enum conversion logic to cover this scenario as well and avoid errors that are hard to understand for users.

Added also a test for it

cc @amyeroberts

### CODE REVIEW

**Review by amyeroberts — COMMENTED** [review]
Thanks for opening this PR! 

I'm not sure why we need this logic. AWQLinearVersion was just added 5 days ago to the repo, and there wasn't a version when this was upper case on `main`, so there's no backwards compatibility for us to handle. Anything which has e.g. "GEMM" was working on a development branch which we don't guarantee to support. The model(s) on the hub should be updated to reflect the approved design.

**Comment by younesbelkada** [comment]
@amyeroberts yes correct there is no backward compatibility to handle, it is more about users that can easily make the typo ("GEMM" vs "gemm") and lead to an error that is hard to understand for them. 
For quantizing a model one needs to call (through autoawq package):

```python
from awq import AutoAWQForCausalLM
from transformers import AutoTokenizer

model_path = "facebook/opt-125m"
quant_path = "opt-125m-awq"
quant_config = {"zero_point": True, "q_group_size": 128, "w_bit": 4, "version":"GEMM"}

# Load model
model = AutoAWQForCausalLM.from_pretrained(model_path)
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)

# Quantize
model.quantize(tokenizer, quant_config=quant_config)
```

Then convert `quanti_config` to the transformers format like this:

```python
from transformers import AwqConfig, AutoConfig
from huggingface_hub import HfApi

# modify the config file so that it is compatible with transformers integration
quantization_config = AwqConfig(
    bits=quant_config["w_bit"],
    group_size=quant_config["q_group_size"],
    zero_point=quant_config["zero_point"],
    version=quant_config["version"].lower(),
).to_dict()

# the pretrained transformers model is stored in the model attribute + we need to pass a dict
model.model.config.quantization_config = quantization_config
# a second solution would be to use Autoconfig and push to hub (what we do at llm-awq)


# save model weights
model.save_quantized(quant_path)
tokenizer.save_pretrained(quant_path)
```

note the `version=quant_config["version"].lower()` here. This PR is more about the user experience and smooth compatibility between autoawq and transformers.

**Comment by amyeroberts** [comment]
@younesbelkada Just to make sure I've understood, from the example above it seems that it's necessary for the value to be uppercase when creating the quant config? 

i.e. `quant_config = {"version": "GEMM"}` is correct but `quant_config = {"version": "gemm"}` is not?

**Comment by younesbelkada** [comment]
@amyeroberts when using external tools such as AutoAWQ or LLm-awq to quantize your LLM yes! However currently users need to make it lowercase before saving it locally / pushing it on the Hub in order to use it with transformers

File: src/transformers/utils/quantization_config.py [hunk 1]
```diff
@@ -567,9 +577,7 @@ def post_init(self):
             )
 
         if self.version not in [AWQLinearVersion.GEMM, AWQLinearVersion.GEMV]:
```
Comment (amyeroberts) [inline src/transformers/utils/quantization_config.py:579]: These are actually doing two different checks. One is checking that its supported in the logic of this class. The other (in the enum) is just checking the enum value. We should do the conversion to the lowercase enum first and then check if it's supported. 

```suggestion
        self.version = AWQLinearVersion.from_str(self.version)
        if self.version not in [AWQLinearVersion.GEMM, AWQLinearVersion.GEMV]:
            raise ValueError(
                f"Only supported versions are in [AWQLinearVersion.GEMM, AWQLinearVersion.GEMV] - not recognized version {self.version}"
            )
```

File: src/transformers/utils/quantization_config.py [hunk 2]
```diff
@@ -567,9 +577,7 @@ def post_init(self):
...
-            raise ValueError(
-                f"Only supported versions are in [AWQLinearVersion.GEMM, AWQLinearVersion.GEMV] - not recognized version {self.version}"
-            )
+            self.version = AWQLinearVersion.from_str(self.version)
```
Comment (amyeroberts) [inline src/transformers/utils/quantization_config.py:580]: Other change to be inline with above (I can't make a suggestion across additions and deletions in the diff)
```suggestion
```

**Review by amyeroberts — COMMENTED** [review]
@younesbelkada OK, thanks for clarifying. I realise it was on my request to use lower case values in the enum. For future PRs it would be preferable to keep alignment (if possible) on the initial integration implementation and I'm very happy to have push back on a request for this reason. 

As a general comment, [ExplicitEnum](https://github.com/huggingface/transformers/blob/bd8f45b1674be4da875097cb5557557d80bcce46/src/transformers/utils/generic.py#L448) was added to provide clearer error  messages for enums.

File: src/transformers/utils/quantization_config.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/utils/quantization_config.py:579]: Makes sense!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing and iterating!

**Comment by amyeroberts** [comment]
@younesbelkada Failing CI runs should now be resolved on main. You should be able to rebase and merge :)

**Comment by amyeroberts** [comment]
@younesbelkada - the test should (🤞 ) now be passing. A full run of all the tests with a fix / skipping tests has passed and been merged to main

**Comment by younesbelkada** [comment]
Thank you veyr much @amyeroberts @ydshieh @muellerzr for fixing the CI issues!

**Pull request merged**